### PR TITLE
Partial renderer: take in account change in clipping and parent geometry

### PR DESCRIPTION
**Pull request by ogoffart**
... as well as opacity

When an item change its geometry or clipping, one really must also refresh all the children, even the ones outside of the item.

And if the position of the parent changes, the old offset must still be used to compute the geometry

This change also add the clip to the dirty region so we don't need to draw so much for clipped items

### CODE REVIEW

**Comment by ogoffart** [comment]
This fixes most of the artifacts in the energy-monitor demo.  
But there is still some artifact remaining as we don't make the region that were covered by a `if` as dirty when the condition becomes false.

Should also fix #2261

**Comment by ogoffart** [comment]
Unfortunately there is no tests yet for the partial renderer. :-/

File: internal/core/item_rendering.rs [hunk 1]
```diff
@@ -406,30 +416,84 @@ impl<'a, T> PartialRenderer<'a, T> {
...
                             });
-                            self.mark_dirty_rect(old_geom, *offset);
-                            self.mark_dirty_rect(geom, *offset);
-                            ItemVisitorResult::Continue(*offset + geom.origin.to_vector())
+
+                            self.mark_dirty_rect(old_geom, state.old_offset, &state.clipped);
+                            self.mark_dirty_rect(geom, state.offset, &state.clipped);
+
+                            new_state.offset += geom.origin.to_vector();
+                            new_state.offset += geom.origin.to_vector();
```
Comment (tronical) [inline internal/core/item_rendering.rs:424]: Did you mean `old_offset` here?:)

Comment (ogoffart) [reply internal/core/item_rendering.rs:424]: Good catch!

File: internal/core/item_rendering.rs [hunk 2]
```diff
@@ -406,30 +416,84 @@ impl<'a, T> PartialRenderer<'a, T> {
...
+                                self.mark_dirty_rect(*data, state.offset, &state.clipped);
+                            }
+
+                            new_state.offset += data.origin.to_vector();
+                            new_state.old_offset += data.origin.to_vector();
+                            if crate::properties::evaluate_no_tracking(|| is_clipping_item(item)) {
+                                new_state.clipped = new_state
+                                    .clipped
+                                    .intersection(
+                                        &data
```
Comment (tronical) [inline internal/core/item_rendering.rs:450]: Might be worth to rename `data` to `cached_item_rect`

File: internal/core/item_rendering.rs [hunk 3]
```diff
@@ -393,11 +393,21 @@ impl<'a, T> PartialRenderer<'a, T> {
...
         crate::item_tree::visit_items(
             component,
             crate::item_tree::TraversalOrder::BackToFront,
-            |_, item, _, offset| {
+            |_, item, _, state| {
+                let mut new_state = *state;
                 let mut borrowed = self.cache.borrow_mut();
+
                 match item.cached_rendering_data_offset().get_entry(&mut borrowed) {
                     Some(CachedGraphicsData { data, dependency_tracker: Some(tr) }) => {
```
Comment (tronical) [inline internal/core/item_rendering.rs:412]:
```suggestion
                    Some(CachedGraphicsData { data: cached_item_rect, dependency_tracker: Some(tr) }) => {
```

**Review by tronical — APPROVED** [review]

**Pull request merged**